### PR TITLE
#463, #466, #250, #448: Treasure Rooms

### LINKED ISSUE #250: Room Type: Treasure
Allows players to pick a treasure type (leader's pick?).

### LINKED ISSUE #448: Mechabees should subtract their bounty from the running total when the self-destruct

### LINKED ISSUE #463: equipment tier obsoleted by Labyrinths
```
		this.tier = tierInput; //TODO obsoleted by Labyrinths
```

https://github.com/Imaginary-Horizons-Productions/prophets-of-the-labyrinth/blob/f2274611d85a09815aeea5833547f5aae1eb98c9/Classes/EquipmentTemplate.js#L5

### LINKED ISSUE #466: add rooms to Labyrinth
```
	// avaialableRooms = []; //TODO add rooms to Labyrinth
```

undefined

### PR DESCRIPTION
**Pull request by ntseng**
Summary
-------
THE LAST MEGA-PR, I PROMISE
- removed vestiges of EquipmentTemplate.tier
- added room declaration to Labyrinth
- split room generation in to resource generation (roll loot/options) and message rendering (generate message and components from resources)
- merged saleList and resourceList properties of room templates by changing shape to an Array of objects to hold more property data
- moved updateRoomHeader and roomHeaderString into roomDAO
- roomTypes not tracked in labyrinth instead of room templates

Local Tests Performed
---------------------
- Tested message renderer for the following cases
   - Battle
   - Events
   - Merchants
   - Victory
   - Give-up
   - Treasure
- Tried each of the treasure room loot types
- Collected loot after battle
- Advanced to new rooms

Remaining Bugs in PR
------------------------
- [ ] complete adventure AFTER opportunity to collect artifact
- [x] "decide the next room on each combat phase" is not yet conditional
- [x] room actions don't decrement down on the room card
- [x] multiple completion embeds on completion

Issue
-----
Closes #250
Closes #463
Closes #466 
Closes #448

### CODE REVIEW

File: Source/Rooms/treasure-basic.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+const ResourceTemplate = require("../../Classes/ResourceTemplate.js");
+const RoomTemplate = require("../../Classes/RoomTemplate.js")
+
+module.exports = new RoomTemplate("Treasure!", [
```
Comment (doubledh) [inline Source/Rooms/treasure-basic.js:4]: can I interpret this combo of loot as: "Some of everything is up for grabs"?

File: Source/Rooms/event-hpshare.js [hunk 2]
```diff
@@ -1,16 +1,13 @@
 const { MessageActionRow, MessageButton } = require("discord.js");
 const RoomTemplate = require("../../Classes/RoomTemplate.js")
 
-module.exports = new RoomTemplate()
-	.setTypes("Event")
-	.setTitle("Health Redistrabution")
-	.setDescription("An imp wearing glasses approaches you with a contract. It would allow you to heal your party members at the expense of the life of one of your own.")
+module.exports = new RoomTemplate("Health Redistrabution", [
```
Comment (doubledh) [inline Source/Rooms/event-hpshare.js]: typo redistrabution -> redistribution
(but also that's the room title so it's probably linked....)

File: Classes/Resource.js [hunk 3]
```diff
@@ -1,27 +1,18 @@
-// Read and write object for adventure.room.resources
 module.exports = class Resource {
-	/**
+	/** This read-write payload class describes a single resource available in an adventure's room
 	 * @param {string} nameInput Note: all names in the combined pool of equipment, artifacts, consumables, and resources must be unique
-	 * @param {"equipment" | "artifact" | "gold" | "scouting" | "roomActions" | "challenge"} resourceTypeInput
+	 * @param {"equipment" | "artifact" | "gold" | "scouting" | "roomAction" | "challenge"| "consumable"} resourceTypeInput
 	 * @param {number} countInput
-	 * @param {"loot" | "merchant" | "resource"} uiTypeInput
+	 * @param {"loot" | "always" | "internal"} visibilityInput
```
Comment (doubledh) [inline Classes/Resource.js]: can I have an explanation (in comments somewhere) for what these different visibilities are?

File: Source/roomDAO.js [hunk 4]
```diff
@@ -1,15 +1,144 @@
...
+ * @param {MessageEmbed} embed
+ * @param {Adventure} adventure
+ */
+function addScoreField(embed, adventure) {
+	const isSuccess = adventure.lives > 0 && adventure.depth === getLabyrinthProperty(adventure.labyrinth, "maxDepth");
+	const livesScore = adventure.lives * 10;
+	const goldScore = Math.floor(Math.log10(adventure.peakGold)) * 5;
+	let score = adventure.accumulatedScore + livesScore + goldScore + adventure.depth;
+	let challengeMultiplier = 1;
+	Object.keys(adventure.challenges).forEach(challengeName => {
```
Comment (doubledh) [inline Source/roomDAO.js:106]: should this be "challenges" or "challenge"?

File: Source/Rooms/event-elementswap.js [hunk 5]
```diff
@@ -1,10 +1,8 @@
 const { MessageActionRow, MessageButton } = require("discord.js");
 const RoomTemplate = require("../../Classes/RoomTemplate.js")
 
-module.exports = new RoomTemplate()
-	.setTypes("Event")
-	.setTitle("Element Redistrabution")
```
Comment (doubledh) [inline Source/Rooms/event-elementswap.js:6]: typo redistrabution -> redistribution
(linked room name)

File: Source/Rooms/event-freegoldonfire.js [hunk 6]
```diff
@@ -1,17 +1,15 @@
...
+const ResourceTemplate = require("../../Classes/ResourceTemplate.js");
 const RoomTemplate = require("../../Classes/RoomTemplate.js")
 
-module.exports = new RoomTemplate()
-	.setTypes("Event")
-	.setTitle("Free Gold?")
-	.setDescription("A large pile of gold sits quietly in the middle of the room, seemingly alone.")
+module.exports = new RoomTemplate("Free Gold?", [
+	new ResourceTemplate("gold", "50", "internal")
+]).setDescription("A large pile of gold sits quietly in the middle of the room, seemingly alone.")
```
Comment (doubledh) [inline Source/Rooms/event-freegoldonfire.js:7]: going to test

File: Classes/ResourceTemplate.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+module.exports = class ResourceTemplate {
+	/** This read-only data class that defines resources available for placement in rooms
+	 * @param {"equipment" | "artifact" | "gold" | "scouting" | "roomAction" | "challenge" | "consumable"} resourceTypeInput
+	 * @param {string} countExpression
+	 * @param {"loot" | "always" | "internal"} visibilityInput
```
Comment (doubledh) [inline Classes/ResourceTemplate.js]: can I have an explanation (in comments somewhere) for what these different visibilities are?

File: Blueprints/labyrinth_blueprint.js [hunk 8]
```diff
@@ -146,6 +146,16 @@ module.exports = new Labyrinth("name", "Untyped", 10)
 				]
 			}
 		})
-	// .setRooms([
-
-	// ]);
+	.setRooms(
+		{
+			"Event": ["Element Redistrabution", "Free Gold?", "Health Redistrabution", "The Score Beggar", "Abandoned Forge", "Equipment Merchant", "Rest Site", "Treasure!"],
```
Comment (doubledh) [inline Blueprints/labyrinth_blueprint.js]: are the merchant/forge/rest/treasure supposed to be both events and their own roomtypes?

**Review by doubledh — COMMENTED** [review]
sttill reading....

File: Source/Rooms/treasure-basic.js (hunk 1 above)
Comment (ntseng) [reply Source/Rooms/treasure-basic.js:4]: "One category of your choice" due to taking stuff costing room actions. Admittedly, theming is sorta weird, but I wanted players to discuss what kinds of treasure they wanted in this room.

File: Source/Rooms/event-freegoldonfire.js (hunk 6 above)
Comment (ntseng) [reply Source/Rooms/event-freegoldonfire.js:7]: Testing if getting killed by the burning gold ends the game properly? Worked ok in my tests.

File: Blueprints/labyrinth_blueprint.js (hunk 8 above)
Comment (ntseng) [reply Blueprints/labyrinth_blueprint.js]: Yep!

File: Source/Buttons/replaceequipmenttreasure.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	(interaction, [name, index, atMerchant]) => {
+		const adventure = getAdventure(interaction.channel.id);
+		if (adventure?.room.roomAction.count > 0) {
+			const { count, cost } = adventure.room.resources[name];
+			if (count > 0) {
+				let delver = adventure.delvers.find(delver => delver.id === interaction.user.id);
+				let discardedName = delver.equipment[index].name;
+				delver.equipment.splice(index, 1, { name, uses: getEquipmentProperty(name, "maxUses") });
+				interaction.channel.messages.fetch(adventure.messageIds.room).then(roomMessage => {
+					adventure.room.resources[name].count--;
```
Comment (doubledh) [inline Source/Buttons/replaceequipmenttreasure.js:19]: pretty sure there's a race condition between line 14 and here (there's a .then on 18)

File: Source/labyrinths/_labyrinthDictionary.js [hunk 10]
```diff
@@ -51,3 +53,41 @@ exports.rollEquipmentDrop = function (tier, adventure) {
...
+exports.rollRoom = function (type, adventure) {
+	if (type === "Artifact Guardian") {
+		return getRoom(adventure.artifactGuardians[adventure.scouting.artifactGuardiansEncountered]);
+	}
+
+	if (type === "Final Battle") {
+		return getRoom(adventure.finalBoss);
+	}
+
+	if (type in LABYRINTHS[adventure.labyrinth]) {
```
Comment (doubledh) [inline Source/labyrinths/_labyrinthDictionary.js:85]: Should this be "type not in"?

File: Source/roomDAO.js [hunk 11]
```diff
@@ -1,15 +1,144 @@
...
+						.setDisabled(!Object.values(adventure.consumables).some(quantity => quantity > 0))
+				));
+			}
+			if (roomTemplate) {
+				components.push(...roomTemplate.uiRows);
+			}
+			components.push(...exports.generateMerchantRows(adventure));
+			components = components.slice(0, MAX_MESSAGE_ACTION_ROWS - 2);
+			if (isCombatVictory) {
+				components.push(exports.generateLootRow(adventure));
```
Comment (doubledh) [inline Source/roomDAO.js]: possible race condition here... maybe?
I get the feeling someone might be able to re-render to duplicate loot as they take it.... well, if they could trigger re-rendering on their own at least (perhaps with a slash command if they get stuck)

Eh, the generated-loot-row isn't the actual resource on the room though, so it should be gated from race conditions by that resource system

**Review by doubledh — COMMENTED** [review]
going to test a couple times...

**Comment by doubledh** [comment]
bug during scouting guardian
```
The merchant reveals that the 1st artifact guardian for this adventure will be undefined (you can review this with /party-stats).
``` 
   followed by this crash  when trying to route to guardian
```
TypeError: Cannot destructure property 'title' of 'undefined' as it is undefined.
    at new Room (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Classes\Room.js:8:23)
    at Object.exports.nextRoom (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Source\adventureDAO.js:123:19)
    at exports.endRoom (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Source\adventureDAO.js:213:10)
    at C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Source\Buttons\routevote.js:42:6
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    ```

File: Source/Rooms/artifactguardian-treasureelemental.js [hunk 12]
```diff
@@ -1,8 +1,8 @@
+const ResourceTemplate = require("../../Classes/ResourceTemplate.js");
 const RoomTemplate = require("../../Classes/RoomTemplate.js")
 
-module.exports = new RoomTemplate()
-	.setTypes("Artifact Guardian")
-	.setTitle("A windfall of treasure!")
-	.setDescription("Floor to ceiling, gold coins, gems and other valuables are stacked in massive piles. Out of the corner of your eyes, you notice a mass of treasure meld together...")
+module.exports = new RoomTemplate("A windfall of treasure!", [
```
Comment (doubledh) [inline Source/Rooms/artifactguardian-treasureelemental.js:4]: consistency: Titlecase on room names?

**Comment by doubledh** [comment]
> bug during scouting guardian
> 
> ```
> The merchant reveals that the 1st artifact guardian for this adventure will be undefined (you can review this with /party-stats).
> ```
> 
> followed by this crash when trying to route to guardian
> 
> ```
> TypeError: Cannot destructure property 'title' of 'undefined' as it is undefined.
>     at new Room (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Classes\Room.js:8:23)
>     at Object.exports.nextRoom (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Source\adventureDAO.js:123:19)
>     at exports.endRoom (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Source\adventureDAO.js:213:10)
>     at C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Source\Buttons\routevote.js:42:6
>     at processTicksAndRejections (node:internal/process/task_queues:96:5)
>     ```
> ```

[{"id":"1041180648171393115","name":"Shining Maze of Earth","labyrinth":"Debug Dungeon","state":"completed","element":"Earth","messageIds":{"utility":"1041180805180969032","room":"1041180982058942464","battleRound":"1041180807219392552"},"leaderId":"113108081990176768","delvers":[{"title":"Chemist","hp":300,"maxHp":300,"block":0,"speed":100,"roundSpeed":7,"actionSpeed":0,"crit":false,"critNumerator":1,"critDenominator":4,"element":"Water","modifiers":{},"staggerThreshold":3,"name":"doubledh","team":"delver","predict":"Health","equipment":[{"name":"Sickle","uses":9},{"name":"Potion Kit","uses":10}],"startingArtifact":"Phoenix Fruit Blossom","id":"113108081990176768","adventureId":"1041180648171393115"}],"challenges":{"Can't Hold All this Value":{"intensity":1,"duration":null},"Restless":{"intensity":30,"duration":null}},"scouting":{"finalBoss":false,"artifactGuardians":1,"artifactGuardiansEncountered":0},"finalBoss":"","artifactGuardians":[null,null],"accumulatedScore":37,"depth":2,"room":{"resources":{"Potion Kit":{"name":"Potion Kit","resourceType":"equipment","visibility":"always","count":1,"cost":200,"uiGroup":"equipment→?"},"Sweeping Ice Ward":{"name":"Sweeping Ice Ward","resourceType":"equipment","visibility":"always","count":1,"cost":350,"uiGroup":"equipment→Rare"},"bossScouting":{"name":"bossScouting","resourceType":"scouting","visibility":"always","count":true,"cost":150,"uiGroup":"scouting"},"guardScouting":{"name":"guardScouting","resourceType":"scouting","visibility":"always","count":true,"cost":100,"uiGroup":"scouting"}},"title":"Equipment Merchant","element":"Earth","hasEnemies":false},"roomCandidates":{"Forge→2":[],"Artifact Guardian→2":[]},"lives":3,"gold":0,"peakGold":100,"artifacts":{"Phoenix Fruit Blossom":{"count":1}},"consumables":{},"rnIndices":{"general":17,"battle":9},"rnTable":"a85099952458894055693847129891125681bb852540281a0144131638511168137561b58b","initialSeed":"1668307210440","guildId":"393489099731763230"},{"id":"1041182163892179025","name":"Shining Maze of Earth","labyrinth":"Debug Dungeon","state":"ongoing","element":"Earth","messageIds":{"utility":"1041182297531088966","room":"1041182358570795048","battleRound":"1041182299456274524"},"leaderId":"113108081990176768","delvers":[{"title":"Chemist","hp":300,"maxHp":300,"block":0,"speed":100,"roundSpeed":7,"actionSpeed":0,"crit":false,"critNumerator":1,"critDenominator":4,"element":"Water","modifiers":{},"staggerThreshold":3,"name":"doubledh","team":"delver","predict":"Health","equipment":[{"name":"Sickle","uses":9},{"name":"Potion Kit","uses":10}],"startingArtifact":"Phoenix Fruit Blossom","id":"113108081990176768","adventureId":"1041182163892179025"}],"challenges":{"Can't Hold All this Value":{"intensity":1,"duration":null},"Restless":{"intensity":30,"duration":null}},"scouting":{"finalBoss":false,"artifactGuardians":1,"artifactGuardiansEncountered":0},"artifactGuardians":[null,null],"accumulatedScore":0,"depth":2,"room":{"resources":{"Potion Kit":{"name":"Potion Kit","resourceType":"equipment","visibility":"always","count":1,"cost":200,"uiGroup":"equipment→?"},"Sweeping Ice Ward":{"name":"Sweeping Ice Ward","resourceType":"equipment","visibility":"always","count":1,"cost":350,"uiGroup":"equipment→Rare"},"bossScouting":{"name":"bossScouting","resourceType":"scouting","visibility":"always","count":true,"cost":150,"uiGroup":"scouting"},"guardScouting":{"name":"guardScouting","resourceType":"scouting","visibility":"always","count":true,"cost":100,"uiGroup":"scouting"}},"title":"Equipment Merchant","element":"Earth","hasEnemies":false},"roomCandidates":{"Forge→2":[],"Artifact Guardian→2":[]},"lives":3,"gold":0,"peakGold":100,"artifacts":{"Phoenix Fruit Blossom":{"count":1}},"consumables":{},"rnIndices":{"general":17,"battle":9},"rnTable":"a85099952458894055693847129891125681bb852540281a0144131638511168137561b58b","initialSeed":"1668307210440","guildId":"393489099731763230"}]


guardians is [null,null], that doesn't seem quite right

**Comment by doubledh** [comment]
seed 1 crash on taking challenge:

```
C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Source\Selects\challenge.js:16
                        if (adventure.room.resources.roomActions.count > 0) {
                                                                 ^

TypeError: Cannot read properties of undefined (reading 'count')
    at Select.execute (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Source\Selects\challenge.js:16:45)
    at exports.callSelect (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\Source\Selects\_selectDictionary.js:22:27)
    at Client.<anonymous> (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\bot.js:108:4)
    at Client.emit (node:events:526:28)
    at InteractionCreateAction.handle (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\node_modules\discord.js\src\client\actions\InteractionCreate.js:83:12)
    at Object.module.exports [as INTERACTION_CREATE] (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\node_modules\discord.js\src\client\websocket\handlers\INTERACTION_CREATE.js:4:36)
    at WebSocketManager.handlePacket (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\node_modules\discord.js\src\client\websocket\WebSocketManager.js:346:31)
    at WebSocketShard.onPacket (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\node_modules\discord.js\src\client\websocket\WebSocketShard.js:478:22)
    at WebSocketShard.onMessage (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\node_modules\discord.js\src\client\websocket\WebSocketShard.js:317:10)
    at WebSocket.onMessage (C:\Users\henry\Documents\dev\prophets-of-the-labyrinth\node_modules\ws\lib\event-target.js:199:18)
    ```

**Comment by doubledh** [comment]
last one must be because "roomActions" (plural) rather than "roomAction"

**Comment by doubledh** [comment]
seed 1 final battle crashed (same error as guardian, no scouting)

**Comment by doubledh** [comment]
forge same room actions bug

**Comment by doubledh** [comment]
I was able to use score beggar 3 times in a row... I think there's a race condition going on there with the room-rendering

**Comment by doubledh** [comment]
doubledh the Chemist gained 62.99999999999999 hp.

(during a rest) haha, floating point math. Probably because of restless challenge?

File: Source/Rooms/event-freegoldonfire.js (hunk 6 above)
Comment (doubledh) [reply Source/Rooms/event-freegoldonfire.js:7]: funny enough I got the gold and didn't take damage AFAICT.... 
That said, I'm pretty sure I was incorrect about predicting this bug, and I think it's race-condition like the score beggar

**Comment by doubledh** [comment]
treasure room doesn't have a select for me,
Combat rooms have "Decide the next room" before combat is over

**Comment by doubledh** [comment]
couldn't find any loot after combats

File: Source/roomDAO.js (hunk 4 above)
Comment (ntseng) [reply Source/roomDAO.js:106]: It's `challenges` in the class. I think that makes more sense, as it's instanced to the adventure "the `challenges` that apply to *this adventure*".

**Comment by ntseng** [comment]
In the absence of being able to resolve comments, I'm gonna use 👍 to mark addressed comments.

File: Source/Rooms/artifactguardian-treasureelemental.js (hunk 12 above)
Comment (ntseng) [reply Source/Rooms/artifactguardian-treasureelemental.js:4]: Haven't thought through the art direction on this on yet. Opinions?

**Comment by ntseng** [comment]
> I was able to use score beggar 3 times in a row... I think there's a race condition going on there with the room-rendering

Haven't been able to reproduce this. Is it still happening for you?

**Review by doubledh — APPROVED** [review]
noooooooice

I think I noticed room actions not updating on card for forge, still, but we can kick that can down

**Pull request merged**